### PR TITLE
Resource Reference Invalidation Improvements

### PR DESCRIPTION
**Pull request by dsainati1**
Closes https://github.com/dapperlabs/cadence-internal/issues/177

This properly invalidates optional resource references, and properly handles the case where an array or dictionary containing an invalid reference is logged or otherwise iterated on. 

______

- [X] Targeted PR against `master` branch
- [X] Linked to Github issue with discussion and accepted design OR link to spec that describes this work
- [X] Code follows the [standards mentioned here](https://github.com/onflow/cadence/blob/master/CONTRIBUTING.md#styleguides)
- [X] Updated relevant documentation 
- [X] Re-reviewed `Files changed` in the Github PR explorer
- [X] Added appropriate labels

### CODE REVIEW

File: runtime/interpreter/interpreter_expression.go [hunk 1]
```diff
@@ -382,6 +382,10 @@ func (interpreter *Interpreter) evalExpression(expression ast.Expression) Value
 }
 
 func (interpreter *Interpreter) checkInvalidatedResourceOrResourceReference(value Value, hasPosition ast.HasPosition) {
+	if interpreter == nil {
+		return
+	}
```
Comment (turbolent) [inline runtime/interpreter/interpreter_expression.go:387]: Why would the receiver be `nil` here?

File: runtime/interpreter/value.go [hunk 2]
```diff
@@ -1940,6 +1940,7 @@ func (v *ArrayValue) iterate(
 			// atree.Array iteration provides low-level atree.Value,
 			// convert to high-level interpreter.Value
 			elementValue := MustConvertStoredValue(interpreter, element)
+			interpreter.checkInvalidatedResourceOrResourceReference(elementValue, locationRange)
```
Comment (turbolent) [inline runtime/interpreter/value.go:1943]: Is adding the calls to `checkInvalidatedResourceOrResourceReference` in the iteration functions (here for arrays, and below for dictionaries and composites) the main fix?

**Comment by turbolent** [comment]
Nice!

File: runtime/interpreter/value.go (hunk 2 above)
Comment (dsainati1) [reply runtime/interpreter/value.go:1943]: Yes, essentially

File: runtime/interpreter/interpreter_expression.go (hunk 1 above)
Comment (dsainati1) [reply runtime/interpreter/interpreter_expression.go:387]: There's a very specific possible sequence of calls that can get here: if `checkInvalidatedResourceOrResourceReference` is called from inside an array/dictionary/composite iteration (as is introduced in this PR), which is triggered by a call to `MeteredString`, which itself is triggered by a call to `RecursiveString` or `String`. While it makes sense to add the full interpreter as an argument to `MeteredString`, requiring the same for calls to `String` or `RecursiveString` is not reasonable, so they use a `nil` interpreter if they delegate to `MeteredString`.

**Review by turbolent — APPROVED** [review]
Great work! 👏

**Review by SupunS — APPROVED** [review]

**Pull request merged**